qemu: nios2: Enable UART 16550 driver for QEMU

ramakrishnapallala opened the pull request:
Enable UART 16550 driver for QEMU as upstream QEMU
does not support JTAG UART.

Jira : RTOS-2362
Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

nashif commented:
note: this required new SDK, do not merge until new SDK is enabled in CI

galak commented on arch/nios2/soc/nios2-qemu/soc.h:
```diff
@@ -0,0 +1,17 @@
+#define UART_NS16550_PORT_0_BASE_ADDR		A_16550_UART_0_BASE
+#define UART_NS16550_PORT_0_IRQ			A_16550_UART_0_IRQ
+#define UART_NS16550_PORT_0_CLK_FREQ		A_16550_UART_0_FREQ
+
```
is it possible to put these defines in board.h instead?

ramakrishnapallala replied:
This change is in line with folder or file structure with arch/nios2/soc/nios2f-zephyr/
I just want to make sure both the folders are aligned.

andyross replied:
Unless it's an off-chip UART integrated by the board, MMIO and IRQ stuff is part of the SoC design and thus should be in soc.h, yeah.

(Though of course this is an FPGA and in principle this can surely be changed at synthesis time and we should probably have yet another layer of abstraction if we really want to be pedantic.  We surely don't).

galak replied:
Wasn't clear to me if the UART was concerned part of the SoC or not, if its part of the SoC than I'm fine with this.

andrewboie replied:
This needs to be at the SOC level, it's how the soft-cpu is wired up.

tyrantlizard77 reviewed (approved):
Need to update to the new SDK before this can go in.

nashif reviewed (changes requested):
blocking merge until new SDK is in

Commit:
qemu: nios2: Enable UART 16550 driver for QEMU

Enable UART 16550 driver for QEMU as upstream QEMU
does not support JTAG UART.

Jira : RTOS-2362
Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

tyrantlizard77 reviewed (changes requested):
This patch doesn't work.
With SDK 0.92 rc 2, I get no console output from 'hello world' with or without this patch

ramakrishnapallala commented:
Just re-verified the changes:
[QEMU] CPU: nios2
        cpu->env.reset_addr:            0
        cpu->env.exception_addr:        400020
***** BOOTING ZEPHYR OS v1.9.0-rc2 - BUILD: Sep  7 2017 13:56:12 *****
Hello World! nios2

@andrewboie , i hope you set the ZEPHYR_SDK_INSTALL_DIR path accordingly.

nashif commented:
reopen when we this becomes needed with a new Qemu in place
